pkg/plugins/policies/meshtrace/plugin/v1alpha1/plugin_test.go

```diff
@@ -0,0 +1,317 @@
+	. "github.com/kumahq/kuma/pkg/xds/envoy/listeners"
+	"github.com/kumahq/kuma/pkg/xds/generator"
+)
+
+var _ = Describe("MeshTrace", func() {
+	type testCase struct {
+		resources         []core_xds.Resource
+		singleItemRules   core_xds.SingleItemRules
+		expectedListeners []string
+		expectedClusters  []string
+	}
+	inboundAndOutbound := []core_xds.Resource{
+		{
+			Name:   "inbound",
+			Origin: generator.OriginInbound,
+			Resource: NewListenerBuilder(envoy_common.APIV3).
+				Configure(InboundListener("inbound:127.0.0.1:17777", "127.0.0.1", 17777, core_xds.SocketAddressProtocolTCP)).
+				Configure(FilterChain(NewFilterChainBuilder(envoy_common.APIV3).
+					Configure(HttpConnectionManager("127.0.0.1:17777", false)),
+				)).MustBuild(),
+		}, {
+			Name:   "outbound",
+			Origin: generator.OriginOutbound,
+			Resource: NewListenerBuilder(envoy_common.APIV3).
+				Configure(OutboundListener("outbound:127.0.0.1:27777", "127.0.0.1", 27777, core_xds.SocketAddressProtocolTCP)).
+				Configure(FilterChain(NewFilterChainBuilder(envoy_common.APIV3).
+					Configure(HttpConnectionManager("127.0.0.1:27777", false)),
+				)).MustBuild(),
+		},
+	}
+	DescribeTable("should generate proper Envoy config",
+		func(given testCase) {
+			resources := core_xds.NewResourceSet()
+			for _, res := range given.resources {
+				r := res
+				resources.Add(&r)
+			}
+
+			context := xds_context.Context{}
+			proxy := xds.Proxy{
+				APIVersion: envoy_common.APIV3,
+				Dataplane: &core_mesh.DataplaneResource{
+					Meta: &test_model.ResourceMeta{
+						Mesh: "default",
+						Name: "backend",
+					},
+					Spec: &mesh_proto.Dataplane{
+						Networking: &mesh_proto.Dataplane_Networking{
+							Inbound: []*mesh_proto.Dataplane_Networking_Inbound{
+								{
+									Tags: map[string]string{
+										mesh_proto.ServiceTag: "backend",
+									},
+									Address: "127.0.0.1",
+									Port:    17777,
+								},
+							},
+							Outbound: []*mesh_proto.Dataplane_Networking_Outbound{
+								{
+									Address: "127.0.0.1",
+									Port:    27777,
+									Tags: map[string]string{
+										mesh_proto.ServiceTag: "other-service",
+									},
+								},
+							},
+						},
+					},
+				},
+				Policies: xds.MatchedPolicies{
+					Dynamic: map[core_model.ResourceType]xds.TypedMatchingPolicies{
+						api.MeshTraceType: {
+							Type:            api.MeshTraceType,
+							SingleItemRules: given.singleItemRules,
+						},
+					},
+				},
+			}
+			plugin := plugin.NewPlugin().(core_plugins.PolicyPlugin)
+
+			Expect(plugin.Apply(resources, context, &proxy)).To(Succeed())
+
+			for i, r := range resources.ListOf(envoy_resource.ListenerType) {
+				actual, err := util_proto.ToYAML(r.Resource)
+				Expect(err).ToNot(HaveOccurred())
+
+				Expect(actual).To(MatchYAML(given.expectedListeners[i]))
+			}
+			Expect(len(resources.ListOf(envoy_resource.ListenerType))).To(Equal(len(given.expectedListeners)))
```

Comment:
This could be moved out to a util package as well. Also this checks the length as well and that does not happen in MeshAccessLog so if there are no resources the test will pass even though it should not. I will fix this in a separate PR.